### PR TITLE
Network recording

### PR DESCRIPTION
**Pull request by mariovisic**
Implement the same tracing that we have in the ruby BK analytics library and use the `request-spy` library to hook into network requests. I have tested this with the axios library and it seems to hook into that as well, so hopefully this will work for most/all? of the JS request libraries available.

~This is almost ready to go, just needs a fix to accomodate how the jest reporters work. The `onTestStart` callback is fired for every test file, not every individual test, so we'll need to hook into each individual test so that we can start with a fresh tracer each time to not double up results~

**This is now ready to go**

This is how it looks on analytics 🙌 

Screen Shot 2022-05-19 at 9 15 26 am

#### TODO:
- [ ] Debug intermittent HTTP reporting
- [ ] Add documentation to the example test suite
- [ ] Update example test suite to confirm to Tim's changes in #22

### CODE REVIEW

**Comment by toolmantim** [comment]
Exciting!

**Review by toolmantim — COMMENTED** [review]
Thanks so much @mariovisic! 🎉

request-spy doesn't seem like it's had much active development (and uses [monkeypatch](https://www.npmjs.com/package/monkeypatch) which seems even older). Did you investigate using Jest’s built-in `spyOn`? Or would that clash with people's own spies/mocks in their tests?

**Comment by mariovisic** [comment]
> request-spy doesn't seem like it's had much active development (and uses [monkeypatch](https://www.npmjs.com/package/monkeypatch) which seems even older). Did you investigate using Jest’s built-in `spyOn`? Or would that clash with people's own spies/mocks in their tests?

I think because those libraries are tiny, means they haven't had to change, [monkeypatch is 25 lines total](https://github.com/robertklep/node-monkeypatch/blob/master/index.js) and [request-spy is 56 lines total](https://github.com/daniel-lundin/request-spy/blob/master/index.js) so they're very simple. If we need to change them we could copy them into the project as they're both just a single file.

Using `spyOn` is a good idea and calls through to the underlying function, so it wouldn't interfere with peoples own spies/mocks, but unfortunately in a jest reporter very oddly there's no "before test" hook and I can't seem to access the `jest` object to call `spyOn`, so unsure if it's possible. It could be done if we get the user to setup the library as a 'preset' in their jest config, but it would likely end up looking very similar to the request-spy code from the library.

**Comment by toolmantim** [comment]
Alright, sounds like a plan just to go with these deps 👍🏼

The example you have in the screenshot, reckon you could check it in under an `examples` directory so I can do a quick test?

**Comment by mariovisic** [comment]
@toolmantim easy, have added it in the examples folder

I've just noticed now that every other test run isn't actually reporting the HTTP times, seems like it could be an async issues so i'll need to address that before this can be merged

**Comment by mariovisic** [comment]
@toolmantim all good rebased now. FYI using the path in the README for the reporter: `'buildkite-collector/jest/reporter'` didn't work for me and I had to use `'buildkite-collector/src/jest/reporter'`

OK, I'm able to reproduce the intermittent HTTP reporting problem, it seems to be how i'm using `onTestCaseResult` as a pseudo before each test hook as there isn't a native one for a jest reporter. If I log some output before running the "before each test" code then it works fine, so seems like a timing issue, I can't think of a nice way to solve that problem however and it's maybe a sign to avoid using `onTestCaseResult` like this.

One solution might be to remove all hook setup from the reporter and have the user also define: `preset: 'buildkite-collector'` in their jest config, that way we can define our own `setupFiles` to create before hooks, as `jest` will be available we can try the `spyOn` approach to see if that also works.

... i'll give that a crack now 👍

**Comment by toolmantim** [comment]
@mariovisic oh, super interesting! Yeah, I don't think having to also define a `preset` is a problem if it gives us what we want. Let me know how that goes.

Good catch on `buildkite-collector/src/jest/reporter` — I did a manual test, but then obviously changed something again. Most probably will have to move `jest` and `util` back to top-level folders.

File: src/util/network.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+const requestSpy = require('request-spy')
+
+class Network {
```
Comment (toolmantim) [inline src/util/network.js:3]: Perhaps we call this `NetworkSpy`?

**Comment by toolmantim** [comment]
Just to follow up on the idea of using `spyOn` — there's a good argument that using `request-spy` is agnostic to the test runner, so we'll be able to use that code across all collectors. I'd suggest don't bother going down that rabbit hole, and we can merge this in first instead.

FYI #22 I added an e2e using the debug output and an example — I imagine this example could follow the same pattern, but be called `jest-http-tracing`

**Comment by mariovisic** [comment]
🤔 I moved to using the `BeforeEach` setup and it's working ok, but i'm still getting that intermittnet problem where it's not reporting HTTP calls sometimes, it seems reproducable now though so will hopefully have some luck tackling that now 👍

**Comment by mariovisic** [comment]
I'm able to replicate the problem now which is a great step forward!

When I limit jest workers to 1 then the problem is fixed, the issue seems to be related to multiple jest workers trying to get/set `global.buildkiteTracer`. To resolve the problem there doesn't seem to be thread local variables I can use, and while I can detect which jest worker is running with: `process.env.JEST_WORKER_ID` in the setup scripts, the reporter hooks don't have `JEST_WORKER_ID` defined (maybe it's run in the parent process), so I don't know which workers test we are reporting on and so can't lookup the correct tracer.

If I could somehow attach the tracer to a property in the test object and then retreive the tracer when the reporter hooks are running, then this problem would be easily solved but i've tried many things and the only way I can effectively "send" values to the reporter from the before script is using the `global` object which is not specific to a single test.

If anyone has any ideas to explore then please let me know ...  for what should be a very straightforward task in jest (run some code before and after the same test in a thread-safe manner) ... i'm not having much luck

**Comment by mariovisic** [comment]
I think we may be able to get proper before/after test hooks working in the reporter by defining a custom test environment similar to what the allure library is doing here: https://github.com/ryparker/jest-circus-allure-environment

Will give that a crack next

**Comment by toolmantim** [comment]
What an adventure @mariovisic! Yeah, it sounds like the allure approach might be the next best bet… looks like you can access all the hooks you need, and store state, using an environment 👍🏼

**Comment by mariovisic** [comment]
#### Bit of an update

I started going down the custom environment approach, but after a few hours of fiddling it seems like a lot of work to be able to inject information into the results, there are some downsides as well like:
  - users can't dynamically change their runner anymore in tests
  - if we only implement a node runner, users can't run e2e tests, to get that working we'll need to maintain another custom environment
  - custom environments only work with jest 24 or later (which is 3 years old, so maybe not a huge deal)

I found this thread where the author wants to do exactly what i'm doing: https://github.com/facebook/jest/issues/7421 but unfortunatly there's no easy options other than implment your own custom environment

So I had a rethink about my assumptions, I think we can solve this problem without resorting to a custom environment, i'll give that a crack later today, hopefully if some of my assumptions are true then it should be simple and easy implement, see how we go :)

**Comment by toolmantim** [comment]
Interesting! I think requiring at Jest >= 24 is probably fine if need be, but those first two downsizes aren't great.

Keen to hear how you go.

**Comment by mariovisic** [comment]
I've figured out why the reporting can be so intermittent, I knew it was due to the jest workers earlier, but I didn't know why jest sometimes seems to allocate multiple workers and sometimes seems to run in a single thread when you don't set any options, there's a GH issue here: https://github.com/facebook/jest/issues/6608 which shows when jest will use multiple workers or not, can be dependent on previous test run time. So i'm now running `jest --clearCache` before each test to get consistent behaviour.

So, i've gone back to using regular hooks and the code actually all works fine when using the `--runInBand` flag on jest, the reason why I think it's working is that the setup code for the HTTP hooks then runs in the same thread as the test runner, as it's all just a single process. When I remove that flag then it doesn't work at all, which makes sense as then jest is using workers and the setup in the reporter isn't happening in the same thread as the test execution.

I don't think setting that flag would be a reasonable solution as it'll dramatically slow down users test suites, so instead the solution should just be to move the network setup to be somewhere that'll run on the jest workes themselves, so `setupFilesAfterEnv` for example, but I haven't been able to figure out a way to send data direct to the reporter, or attach metadata to tests and so can't match up the test to it's tracer so it's not useful.

Other than the custom test environment which I don't think is going to be a very clean solution, i'm a bit stumped, I think i'm just going to leave it here for the day and come back later when I have some more fresh ideas.

**Comment by AeroCross** [comment]
@mariovisic, how did you go? After merging #35, will there be anything left here to investigate or implement further?

**Comment by mariovisic** [comment]
@AeroCross the state of this PR is that the current implementation won't work in jest when multiple worker processes are involved (which is the default), the reporter runs in the parent process and there's no way of sharing information between the reporter and worker processes (which is important to be able to track which http requests occured during which test).

I can't easily find a way to get around this limitation in jest without implementing an entire test environment runtime (similar to what is here): https://github.com/ryparker/jest-circus-allure-environment and I was avoiding implementing a runtime as it seems to come with a lot of complexity and problems of its own.

I've ported the important bits already to #35 and I suspect attempting to implement this feature in the mocha reporter may be much easier. Though make sure to manually test with a test suite with multiple files to ensure there are no worker/async issues like we have in jest

Good luck 👍

**Comment by AeroCross** [comment]
That's a great summary, thank you very much for digging into this 🙏

**Pull request closed without merging**